lib/src/rules/prefer_collection_literals.dart

```diff
@@ -142,35 +142,35 @@ class _Visitor extends SimpleAstVisitor<void> {
 
   bool _shouldSkipLinkedHashLint(
       InstanceCreationExpression node, bool Function(DartType node) typeCheck) {
-    if (_isHashMap(node) || _isHashSet(node)) {
-      // Skip: LinkedHashSet<int> s =  ...; or LinkedHashMap<int> s =  ...;
-      var parent = node.parent;
-      if (parent is VariableDeclaration) {
-        var parent2 = parent.parent;
-        if (parent2 is VariableDeclarationList) {
-          var assignmentType = parent2.type?.type;
-          if (assignmentType != null && !typeCheck(assignmentType)) {
-            return true;
-          }
-        }
-      }
-      // Skip: function(LinkedHashSet()); when function(LinkedHashSet mySet) or
-      // function(LinkedHashMap()); when function(LinkedHashMap myMap)
-      if (parent is ArgumentList) {
-        final paramType = parent.arguments.first.staticParameterElement?.type;
-        if (paramType != null && !typeCheck(paramType)) {
-          return true;
-        }
-      }
-      // Skip: <int, LinkedHashSet>{}.putIfAbsent(3, () => LinkedHashSet());
-      // or <int, LinkedHashMap>{}.putIfAbsent(3, () => LinkedHashMap());
-      if (parent is ExpressionFunctionBody) {
-        var expressionType = parent.expression.staticType;
-        if (expressionType != null && !typeCheck(expressionType)) {
-          return true;
-        }
-      }
+    if (!_isHashMap(node) && !_isHashSet(node)) return false;
+    var parent = node.parent;
+    var staticType = _enforcedType(parent);
+    return staticType != null && !typeCheck(staticType);
+  }
+}
+
+/// Returns the static type which is pushed into an expression by it's parent.
+DartType /*?*/ _enforcedType(AstNode parent) {
+  if (parent is VariableDeclaration) {
+    var parent2 = parent.parent;
+    if (parent2 is VariableDeclarationList) {
+      return parent2.type?.type;
     }
-    return false;
   }
+  if (parent is ArgumentList) {
+    return parent.arguments.first.staticParameterElement?.type;
```

Comment:
The use of `first` here looks suspicious. I'm guessing that you need to pass the `InstanceCreationExpression` into this method rather than its parent, and ask that node for its `staticParameterElement`.

You can test this by defining and using a method like `printIndentedSet(int indent, Set<int> ids)` in the test code. I believe that `first.staticParameterElement` would return an element whose type is `int`.

Reply:
This was the existing behavior, you are correct that it has a bug. Let's fix that separately rather than have it wait to figure out the rest of this PR. https://github.com/dart-lang/linter/pull/2347 should fix it.

@pq - is there a typical process I should be following when fixing these false negatives? I have another category of false negative with this lint that I'd like to fix, but I don't know how much cleanup it will require.

Reply:
Fixing false negatives is a commitment since it will likely lead to breakage on the next  Google3 roll.  Ideally issues get  fixed proactively before a roll is tried.  Happy to share my approach.